pkg/datasource/data.go

```diff
@@ -0,0 +1,429 @@
+			nf.Flags |= uint32(s.FieldFlags())
+		}
+		if s, ok := f.(AnnotatedField); ok {
+			nf.Annotations = s.FieldAnnotations()
+		}
+		if s, ok := f.(ParentedField); ok {
+			parent := s.FieldParent()
+			if parent >= 0 {
+				nf.Parent = uint32(parent + parentOffset) // TODO: validate?
+				nf.Flags |= FieldFlagHasParent.Uint32()
+				parentFields[parent] = struct{}{}
+			}
+		}
+		newFields = append(newFields, nf)
+	}
+
+	// Unref parent fields
+	for p := range parentFields {
+		FieldFlagUnreferenced.AddTo(&newFields[p].Flags)
+	}
+
+	var err error
+	for i, f := range newFields {
+		f.FullName, err = resolveNames(uint32(i), newFields, uint32(parentOffset))
+		if err != nil {
+			return nil, fmt.Errorf("resolving full fieldnames: %w", err)
+		}
+	}
+
+	ds.fields = append(ds.fields, newFields...)
+
+	for _, f := range newFields {
+		ds.fieldMap[f.Name] = f
+	}
+
+	ds.payloadCount++
+
+	return &fieldAccessor{ds: ds, f: &field{
+		PayloadIndex: idx,
+		Size:         size,
+		Flags:        FieldFlagContainer.Uint32(),
+	}}, nil
+}
+
+func (ds *dataSource) AddField(name string, opts ...FieldOption) (FieldAccessor, error) {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	if _, ok := ds.fieldMap[name]; ok {
+		return nil, fmt.Errorf("field %q already exists", name)
+	}
+
+	nf := &field{
+		Name:     name,
+		FullName: name,
+		Index:    uint32(len(ds.fields)),
+		Kind:     api.Kind_Invalid,
+	}
+	for _, opt := range opts {
+		opt(nf)
+	}
+
+	// Reserve new payload for non-empty fields
+	if !FieldFlagEmpty.In(nf.Flags) {
+		nf.PayloadIndex = ds.payloadCount
+		ds.payloadCount++
+	}
+
+	ds.fields = append(ds.fields, nf)
+	ds.fieldMap[nf.FullName] = nf
+	return &fieldAccessor{ds: ds, f: nf}, nil
+}
+
+func (ds *dataSource) GetField(name string) FieldAccessor {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	f, ok := ds.fieldMap[name]
+	if !ok {
+		return nil
+	}
+	return &fieldAccessor{ds: ds, f: f}
+}
+
+func (ds *dataSource) GetFieldsWithTag(tag ...string) []FieldAccessor {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	res := make([]FieldAccessor, 0)
+	for _, f := range ds.fields {
+		for _, t := range tag {
+			if slices.Contains(f.Tags, t) {
+				res = append(res, &fieldAccessor{ds: ds, f: f})
+				break
+			}
+		}
+	}
+	return res
+}
+
+func (ds *dataSource) Subscribe(fn DataFunc, priority int) {
+	if fn == nil {
+		return
+	}
+
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	ds.subscriptions = append(ds.subscriptions, &subscription{
+		priority: priority,
+		fn:       fn,
+	})
+	sort.SliceStable(ds.subscriptions, func(i, j int) bool {
+		return ds.subscriptions[i].priority < ds.subscriptions[j].priority
+	})
+}
+
+func (ds *dataSource) EmitAndRelease(d Data) error {
+	defer ds.dPool.Put(d)
+	for _, sub := range ds.subscriptions {
+		err := sub.fn(ds, d)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ds *dataSource) Release(d Data) {
+	ds.dPool.Put(d)
+}
+
+func (ds *dataSource) ReportLostData(ctr uint64) {
+	// TODO
```

Comment:
tbh, I think this is out of scope (at least for now). Currently, in the run gadget, we're printing a warning, I'm happy to add that as well. Thought behind this is to instead of just sending warnings (potentially increasing load and bandwidth that were responsible for the lost data in the first place), this should later on give you a summary on the client (or at least trigger the textual warnings on the client side) like `8000 events received, 50 dropped`.

However, I'm not sure whether it's a better idea to remove that function or leave it in with the TODO.

Reply:
IMO, you should remove the TODO and create an issue where we can discuss what's the best approach to report what was dropped.